[Parse/AST/Sema] Split parsing for typealias & associatedtype, allow typealias in protocols and generic constraints

gregomni opened the pull request:
Split up parsing of typealias and associatedtype, including dropping a now unneeded ParseDeclOptions flag.

Then made typealias in a protocol actually valid, which works great except that you can't use them in where clauses in generics now. So produce a specific diagnosis if you try to use them as if they were an associatedtype in a where clause, but otherwise they act like you would hope for protocol conformance purposes (i.e. as an alias possibly involved in the types of other func/var conformances, not as a hidden
generic param in itself).

gregomni commented:
This pull is the actual code/feature changes from <https://github.com/apple/swift/pull/1557>, after that pull got revised to just contain `typealias` -> `associatedtype` changes in tests.

gribozavr commented:
@gregomni Awesome!  The standard library team has been waiting for this feature for a long time!

If you want a stress test for your implementation, try adding this to the standard library (we wanted to do this for a log time!):

```swift
extension Collection {
  typealias Element = Iterator.Element
}
```

and replacing `Iterator.Element` with just `Element` in all APIs, protocol extensions, generic parameters, constraints.

gribozavr commented on test/decl/typealias/typealias.swift:
```diff
@@ -122,3 +121,24 @@ extension A<Float,Int> {}  // expected-error {{constrained extension must be dec
+}
+
+// expected-error @+1 {{'CB' typealias 'E' can not be used as a generic constraint; use associatedtype instead}}
+func go1<T : CB, U : Col where U.Elem == T.E>(col: U, builder: T) { // expected-error {{'E' is not a member type of 'T'}}
```
Oh, why can't `E` be used in generic constraints?  Is this a temporary limitation?

gregomni replied:
Yes, it's hopefully just temporary - I'm trying to break these changes into smaller pieces. This pull avoids touching ArchetypeBuilder except for the temporary diagnosis with the limitation.  I have some more work there (stashed at the moment...) that makes generic constraints work for straightforward alias cases (like the `typealias Element = Iterator.Element` that you mentioned), but it isn't quite correct yet. It'll be next up!

gribozavr replied:
I see -- looking forward to it!

gregomni commented:
@gribozavr Added a second commit with the (limited) generic constraint support.

Would love some feedback on whether this seems like a good approach, and whether the intermediate step of parsing but not generically constraining with protocol type aliases ought to be committed separately first. Thanks!

gribozavr commented:
CC @jopamer @cwillmor @DougGregor for archetype builder changes.  Please help to review this PR, it will unblock a massive simplification to the standard library's generic constraints!

benlangmuir commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2850,6 +2798,69 @@ ParserResult<TypeDecl> Parser::parseDeclTypeAlias(bool WantDefinition,
+  ParserStatus Status;
+  
+  Status |=
+  parseIdentifierDeclName(*this, Id, IdLoc, tok::colon, tok::equal,
```
Style: please indent this line 4 spaces since it's part of the `|=`.

benlangmuir commented on include/swift/AST/ArchetypeBuilder.h:
```diff
@@ -466,6 +468,18 @@ class ArchetypeBuilder::PotentialArchetype {
+  : ParentOrParam(Parent), NameOrAssociatedType(TypeAlias),
+  Representative(this), IsRecursive(false), Invalid(false),
+  SubstitutingConcreteType(false), RecursiveConcreteType(false),
+  RecursiveSuperclassType(false), Renamed(false)
```
Style: please indent `:` and align the members (or just run clang-format on it).

jopamer commented:
These look good to me. @DougGregor - any objections to me merging?

DougGregor commented on lib/AST/ArchetypeBuilder.cpp:
```diff
@@ -467,13 +468,46 @@ auto ArchetypeBuilder::PotentialArchetype::getNestedType(
+        // Resolve this nested type to this type alias.
+        pa = new PotentialArchetype(this, alias);
+        
+        if (alias->hasUnderlyingType()) {
```
Shouldn't we call resolveDeclSignature() if the typealias doesn't have an underlying type yet?

DougGregor commented on lib/AST/ArchetypeBuilder.cpp:
```diff
@@ -467,13 +468,46 @@ auto ArchetypeBuilder::PotentialArchetype::getNestedType(
+                  break;
+              }
+              if (!existingPA)
+                continue;
```
In the case where the underlying type does not depend on a type parameter (e.g, !hasArchetype() with the current formulation, or !isDependentType() if this operates on interface types), we could turn this into a same-type requirement.

gregomni replied:
It currently sets pa->ArchetypeOrConcreteType to the concrete type, which appears to work for anywhere this alias appears in a same type requirement. Is there something else I'm missing?

DougGregor commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2798,46 +2770,22 @@ ParserResult<TypeDecl> Parser::parseDeclTypeAlias(bool WantDefinition,
-      assocType->setInherited(Context.AllocateCopy(Inherited));
-    addToScope(assocType);
-    return makeParserResult(Status, assocType);
+  if (Tok.is(tok::colon)) {
```
I'd appreciate if we could improve QoI here for cases where Swift 2-minded developers write "typealias A: P" or "typealias A" when they mean to use "associatedtype". It shouldn't be hard to identify such cases when we're in a protocol context, suggest "associatedtype", and finish processing it as an associated type.

gregomni replied:
Thanks! Did this this morning.

DougGregor commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -174,6 +174,21 @@ Type CompleteGenericTypeResolver::resolveDependentMemberType(
+    auto repr = alias->getUnderlyingTypeLoc().getTypeRepr();
+    if (auto identRepr = dyn_cast<IdentTypeRepr>(repr)) {
+      for (auto component : identRepr->getComponentRange())
+        baseTy = resolveDependentMemberType(baseTy, DC, baseRange, component);
```
Can we use lookupMemberType on the base type here? That way, we should be able to handle arbitrarily complicated typealiases.

gregomni replied:
Can't use lookupMemberType(), unfortunately, as it specifically asserts if you try to call it with a dependent type...
```
  // Callers must cope with dependent types directly.  
  assert(!type->isTypeParameter());
```

DougGregor commented on test/Generics/associated_types.swift:
```diff
@@ -173,7 +173,7 @@ struct C<a : B> : B { // expected-error {{type 'C<a>' does not conform to protoc
 // SR-511
 protocol sr511 {
-  typealias Foo // expected-warning {{use of 'typealias' to declare associated types is deprecated; use 'associatedtype' instead}} {{3-12=associatedtype}}
+  typealias Foo // expected-error {{expected '=' in typealias declaration}}
```
This is a case where we should provide a specialized diagnostic + Fix-It to replace "typealias" with "associatedtype"

DougGregor commented:
Some comments above. The general direction looks very good (thanks for working on this!), but I'd like to avoid the need to walk through IdentTypeReprs... there should be more-general type checker mechanisms to handle these cases.

gregomni commented:
@DougGregor Awesome. Thanks for the detailed feedback!

gregomni commented on lib/AST/ArchetypeBuilder.cpp:
```diff
@@ -467,13 +468,46 @@ auto ArchetypeBuilder::PotentialArchetype::getNestedType(
+            if (auto identRepr = dyn_cast<IdentTypeRepr>(repr)) {
+              PotentialArchetype *existingPA = this;
+              for (auto component : identRepr->getComponentRange()) {
+                existingPA = existingPA->getNestedType(component->getIdentifier(),
```
@DougGregor I can see how to do it the archetype decomposing way, and I can write that code. I'm not sure how interface types work, and it looks like computeInterfaceType() is on NominalDecl, which a TypeAliasDecl isn't. The fallout from changing TypeAliasDecl's superclass is beyond what I'm currently comfortable with. Sound okay to do archetype decomposing, and save interface type for future enhancement?

gregomni replied:
Ignore me here. I was confused. :-) Working on it.

gregomni replied:
Added a commit decomposing the archetypes, definitely better.

gregomni commented:
Added a commit that fixes the indentation issues, better diagnosis for type aliases in protocols that look like they are Swift 2 intended-to-be-associatedtypes, and added tests. Thanks for the feedback @benlangmuir!

gregomni commented:
@DougGregor I think all the feedback has been addressed, tests added, tests pass, etc., except that the added code in TypeCheckGeneric is still using IdentTypeRepr to do its job, and it handles 'Self' in a terrible way. Any suggestions there?

DougGregor commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2797,47 +2769,32 @@ ParserResult<TypeDecl> Parser::parseDeclTypeAlias(bool WantDefinition,
+    // If we're in a protocol and don't see an '=' this might be leftover Swift 2
+    // code intending to be an associatedtype.
+    diagnose(TypeAliasLoc, diag::typealias_inside_protocol_without_type)
+    .fixItReplace(TypeAliasLoc, "associatedtype");
```
Please indent here.

DougGregor commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2850,6 +2807,69 @@ ParserResult<TypeDecl> Parser::parseDeclTypeAlias(bool WantDefinition,
+  ParserResult<TypeRepr> UnderlyingTy;
+  if (Tok.is(tok::equal)) {
+    consumeToken(tok::equal);
+    UnderlyingTy = parseType(diag::expected_type_in_typealias);
```
Do we need an expected_type_in_associatedtype diagnostic?

DougGregor commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -174,6 +174,24 @@ Type CompleteGenericTypeResolver::resolveDependentMemberType(
+  if (auto alias = nestedPA->getTypeAliasDecl()) {
+    if (nestedPA->isConcreteType())
+      return nestedPA->getConcreteType();
+
```
Can you just use TC.substMemberTypeWithBase(dc->getModuleContext(), alias, baseTy, true) here?

DougGregor commented:
Getting very close! Just a few minor comments; per comment about on TypeCheckGeneric, did you try using substMemberTypeWithBase?

gregomni commented:
@DougGregor I think this may be ready now?

DougGregor commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2850,6 +2805,80 @@ ParserResult<TypeDecl> Parser::parseDeclTypeAlias(bool WantDefinition,
+    AssociatedTypeLoc = consumeToken(tok::kw_typealias);
+    diagnose(AssociatedTypeLoc, diag::typealias_inside_protocol_without_type)
+        .fixItReplace(AssociatedTypeLoc, "associatedtype");
+    Status.setIsParseError();
```
You don't need to set Status.setIsParseError(), since you've recovered completely via the Fix-It. That way, we can continue as if the user never made the mistake at all.

DougGregor commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -174,6 +174,16 @@ Type CompleteGenericTypeResolver::resolveDependentMemberType(
+    if (nestedPA->isConcreteType())
+      return nestedPA->getConcreteType();
+
+    return TC.substMemberTypeWithBase(DC->getParentModule(), alias,
```
Fantastic simplification from the previous version.

DougGregor commented:
With that one tweak to remove the unnecessary "setIsParseError()", please merge! I can hear the happiness from Standard Library-land.

gregomni commented:
@DougGregor Cool! Thanks for taking the time to go through all these changes.

gregomni commented:
Removed the setIsParseError, squashed, merging...

gribozavr commented:
@gregomni @DougGregor @jopamer Thank you very much!  :clap:  I'll try it in the library now!

CC @dabrahams @moiseev

dabrahams commented:
W00t!

gribozavr commented:
Sorry @gregomni, I think I broke it: https://bugs.swift.org/browse/SR-937

dabrahams commented:
@DougGregor apparently this isn't quite working (also should be under a flag until reviewed, right?)

gregomni commented:
Yep, it's working in simple cases, but apparently isn't quite right. Working on it here: <https://bugs.swift.org/browse/SR-938>.
